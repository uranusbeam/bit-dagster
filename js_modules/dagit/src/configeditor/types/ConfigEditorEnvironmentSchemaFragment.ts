// @generated
/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: ConfigEditorEnvironmentSchemaFragment
// ====================================================

export interface ConfigEditorEnvironmentSchemaFragment_rootEnvironmentType {
  __typename: "ArrayConfigType" | "CompositeConfigType" | "EnumConfigType" | "NullableConfigType" | "RegularConfigType" | "ScalarUnionConfigType";
  key: string;
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_ArrayConfigType {
  __typename: "ArrayConfigType" | "NullableConfigType";
  key: string;
  description: string | null;
  isSelector: boolean;
  typeParamKeys: string[];
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_RegularConfigType {
  __typename: "RegularConfigType";
  key: string;
  description: string | null;
  isSelector: boolean;
  typeParamKeys: string[];
  givenName: string;
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_EnumConfigType_values {
  __typename: "EnumConfigValue";
  value: string;
  description: string | null;
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_EnumConfigType {
  __typename: "EnumConfigType";
  key: string;
  description: string | null;
  isSelector: boolean;
  typeParamKeys: string[];
  givenName: string;
  values: ConfigEditorEnvironmentSchemaFragment_allConfigTypes_EnumConfigType_values[];
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_CompositeConfigType_fields {
  __typename: "ConfigTypeField";
  name: string;
  description: string | null;
  isOptional: boolean;
  configTypeKey: string;
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_CompositeConfigType {
  __typename: "CompositeConfigType";
  key: string;
  description: string | null;
  isSelector: boolean;
  typeParamKeys: string[];
  fields: ConfigEditorEnvironmentSchemaFragment_allConfigTypes_CompositeConfigType_fields[];
}

export interface ConfigEditorEnvironmentSchemaFragment_allConfigTypes_ScalarUnionConfigType {
  __typename: "ScalarUnionConfigType";
  key: string;
  description: string | null;
  isSelector: boolean;
  typeParamKeys: string[];
  scalarTypeKey: string;
  nonScalarTypeKey: string;
}

export type ConfigEditorEnvironmentSchemaFragment_allConfigTypes = ConfigEditorEnvironmentSchemaFragment_allConfigTypes_ArrayConfigType | ConfigEditorEnvironmentSchemaFragment_allConfigTypes_RegularConfigType | ConfigEditorEnvironmentSchemaFragment_allConfigTypes_EnumConfigType | ConfigEditorEnvironmentSchemaFragment_allConfigTypes_CompositeConfigType | ConfigEditorEnvironmentSchemaFragment_allConfigTypes_ScalarUnionConfigType;

export interface ConfigEditorEnvironmentSchemaFragment {
  __typename: "EnvironmentSchema";
  rootEnvironmentType: ConfigEditorEnvironmentSchemaFragment_rootEnvironmentType;
  allConfigTypes: ConfigEditorEnvironmentSchemaFragment_allConfigTypes[];
}
