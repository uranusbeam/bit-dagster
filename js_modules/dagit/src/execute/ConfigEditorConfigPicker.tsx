import * as React from "react";
import {
  Button,
  Menu,
  Spinner,
  MenuItem,
  Intent,
  IInputGroupProps,
  HTMLInputProps
} from "@blueprintjs/core";
import * as ReactDOM from "react-dom";

import { Select, Suggest } from "@blueprintjs/select";
import { useQuery } from "react-apollo";
import {
  ConfigPresetsQuery,
  ConfigPresetsQuery_pipeline_presets
} from "./types/ConfigPresetsQuery";
import {
  ConfigPartitionsQuery,
  ConfigPartitionsQuery_partitionSetOrError_PartitionSet_partitions_results
} from "./types/ConfigPartitionsQuery";
import {
  ConfigPartitionSetsQuery,
  ConfigPartitionSetsQuery_partitionSetsOrError_PartitionSets_results
} from "./types/ConfigPartitionSetsQuery";
import gql from "graphql-tag";
import { IExecutionSession } from "../LocalStorage";
import { isEqual } from "apollo-utilities";
import styled from "styled-components";
import { ShortcutHandler } from "../ShortcutHandler";

type Preset = ConfigPresetsQuery_pipeline_presets;
type PartitionSet = ConfigPartitionSetsQuery_partitionSetsOrError_PartitionSets_results;
type Partition = ConfigPartitionsQuery_partitionSetOrError_PartitionSet_partitions_results;
type ConfigGenerator = Preset | PartitionSet;
interface Pipeline {
  tags: {
    key: string;
    value: string;
  }[];
}

interface ConfigEditorConfigPickerProps {
  base: IExecutionSession["base"];
  pipelineName: string;
  solidSubset: string[] | null;
  onSaveSession: (updates: Partial<IExecutionSession>) => void;
  onCreateSession: (initial: Partial<IExecutionSession>) => void;
}

const PRESET_PICKER_HINT_TEXT = `Define a PresetDefinition, PartitionSetDefinition, or a schedule decorator (e.g. @daily_schedule) to autofill this session...`;

export class ConfigEditorConfigPicker extends React.Component<
  ConfigEditorConfigPickerProps
> {
  onSelectPartitionSet = (partitionSet: PartitionSet) => {
    this.props.onSaveSession({
      base: {
        partitionsSetName: partitionSet.name,
        partitionName: null
      }
    });
  };

  onSelectPreset = (preset: Preset, pipeline?: Pipeline) => {
    this.onCommit({
      base: { presetName: preset.name },
      name: preset.name,
      environmentConfigYaml: preset.environmentConfigYaml || "",
      solidSubset: preset.solidSubset,
      mode: preset.mode,
      tags: [...(pipeline?.tags || [])]
    });
  };

  onSelectPartition = (partition: Partition, pipeline?: Pipeline) => {
    this.onCommit({
      name: partition.name,
      base: Object.assign({}, this.props.base, {
        partitionName: partition.name
      }),
      environmentConfigYaml: partition.environmentConfigYaml || "",
      solidSubset: partition.solidSubset,
      mode: partition.mode,
      tags: [...(pipeline?.tags || []), ...partition.tags]
    });
  };

  onCommit = (changes: Partial<IExecutionSession>) => {
    this.props.onSaveSession(changes);
  };

  render() {
    const { pipelineName, solidSubset, base } = this.props;

    return (
      <PickerContainer>
        <ConfigEditorConfigGeneratorPicker
          value={base}
          pipelineName={pipelineName}
          solidSubset={solidSubset}
          onSelectPreset={this.onSelectPreset}
          onSelectPartitionSet={this.onSelectPartitionSet}
        />
        {base && "partitionsSetName" in base && (
          <>
            <div style={{ width: 5 }} />
            <ConfigEditorPartitionPicker
              key={base.partitionsSetName}
              pipelineName={pipelineName}
              partitionSetName={base.partitionsSetName}
              value={base.partitionName}
              onSelect={this.onSelectPartition}
            />
          </>
        )}
      </PickerContainer>
    );
  }
}

interface ConfigEditorPartitionPickerProps {
  pipelineName: string;
  partitionSetName: string;
  value: string | null;
  onSelect: (partition: Partition, pipeline?: Pipeline) => void;
}

export const ConfigEditorPartitionPicker: React.FunctionComponent<ConfigEditorPartitionPickerProps> = React.memo(
  props => {
    const { partitionSetName, pipelineName, value, onSelect } = props;
    const { data, loading } = useQuery<ConfigPartitionsQuery>(
      CONFIG_PARTITIONS_QUERY,
      {
        variables: { partitionSetName, pipelineName },
        fetchPolicy: "network-only"
      }
    );

    const partitions: Partition[] =
      data?.partitionSetOrError.__typename === "PartitionSet"
        ? data.partitionSetOrError.partitions.results
        : [];

    const selected = partitions.find(p => p.name === value);

    const inputProps: IInputGroupProps & HTMLInputProps = {
      placeholder: "Partition",
      style: { width: 180 },
      intent: (loading ? !!value : !!selected) ? Intent.NONE : Intent.DANGER
    };

    // If we are loading the partitions and do NOT have any cached data to display,
    // show the component in a loading state with a spinner and fill it with the
    // current partition's name so it doesn't flicker (if one is set already.)
    if (loading && partitions.length === 0) {
      return (
        <Suggest<string>
          key="loading"
          inputProps={{
            ...inputProps,
            rightElement: !value ? <Spinner size={17} /> : undefined
          }}
          items={[]}
          itemRenderer={() => null}
          noResults={<Menu.Item disabled={true} text="Loading..." />}
          inputValueRenderer={str => str}
          selectedItem={value}
        />
      );
    }

    // Note: We don't want this Suggest to be a fully "controlled" React component.
    // Keeping it's state is annoyign and we only want to update our data model on
    // selection change. However, we need to set an initial value (defaultSelectedItem)
    // and ensure it is re-applied to the internal state when it changes (via `key` below).
    return (
      <Suggest<Partition>
        key={selected ? selected.name : "none"}
        defaultSelectedItem={selected}
        items={partitions}
        inputProps={inputProps}
        inputValueRenderer={partition => partition.name}
        itemPredicate={(query, partition) =>
          query.length === 0 || partition.name.includes(query)
        }
        itemRenderer={(partition, props) => (
          <Menu.Item
            active={props.modifiers.active}
            onClick={props.handleClick}
            key={partition.name}
            text={partition.name}
          />
        )}
        noResults={<Menu.Item disabled={true} text="No presets." />}
        onItemSelect={item => onSelect(item, data?.pipeline)}
      />
    );
  },
  isEqual
);

interface ConfigEditorConfigGeneratorPickerProps {
  pipelineName: string;
  solidSubset: string[] | null;
  value: IExecutionSession["base"];
  onSelectPreset: (preset: Preset, pipeline?: Pipeline) => void;
  onSelectPartitionSet: (
    partitionSet: PartitionSet,
    pipeline?: Pipeline
  ) => void;
}

export const ConfigEditorConfigGeneratorPicker: React.FunctionComponent<ConfigEditorConfigGeneratorPickerProps> = React.memo(
  props => {
    const { pipelineName, onSelectPreset, onSelectPartitionSet, value } = props;
    const {
      presets,
      partitionSets,
      loading,
      pipeline
    } = usePresetsAndPartitions(pipelineName);

    const configGenerators: ConfigGenerator[] = [...presets, ...partitionSets];
    const empty = !loading && configGenerators.length === 0;

    const select: React.RefObject<Select<ConfigGenerator>> = React.createRef();
    const onSelect = (item: ConfigGenerator) => {
      if (item.__typename === "PartitionSet") {
        onSelectPartitionSet(item, pipeline);
      } else {
        onSelectPreset(item, pipeline);
      }
    };

    let emptyLabel = `Preset / Partition Set`;
    if (presets.length && !partitionSets.length) {
      emptyLabel = `Preset`;
    } else if (!presets.length && partitionSets.length) {
      emptyLabel = `Partition Set`;
    }

    const label = !value
      ? emptyLabel
      : "presetName" in value
      ? `Preset: ${value.presetName}`
      : `Partition Set: ${value.partitionsSetName}`;

    return (
      <div>
        <ShortcutHandler
          shortcutLabel={"âŒ¥E"}
          shortcutFilter={e => e.keyCode === 69 && e.altKey}
          onShortcut={() => activateSelect(select.current)}
        >
          <Select<ConfigGenerator>
            ref={select}
            disabled={empty}
            items={configGenerators}
            itemPredicate={(query, configGenerator) =>
              query.length === 0 || configGenerator.name.includes(query)
            }
            itemListRenderer={({
              itemsParentRef,
              renderItem,
              filteredItems
            }) => {
              const renderedPresetItems = filteredItems
                .filter(item => item.__typename === "PipelinePreset")
                .map(renderItem)
                .filter(Boolean);

              const renderedPartitionSetItems = filteredItems
                .filter(item => item.__typename === "PartitionSet")
                .map(renderItem)
                .filter(Boolean);

              const bothTypesPresent =
                renderedPresetItems.length > 0 &&
                renderedPartitionSetItems.length > 0;

              return (
                <Menu ulRef={itemsParentRef}>
                  {bothTypesPresent && (
                    <MenuItem disabled={true} text={`Presets`} />
                  )}
                  {renderedPresetItems}
                  {bothTypesPresent && <Menu.Divider />}
                  {bothTypesPresent && (
                    <MenuItem disabled={true} text={`Partition Sets`} />
                  )}
                  {renderedPartitionSetItems}
                </Menu>
              );
            }}
            itemRenderer={(item, props) => (
              <Menu.Item
                active={props.modifiers.active}
                onClick={props.handleClick}
                key={item.name}
                text={
                  <div>
                    {item.name}
                    <div style={{ opacity: 0.4, fontSize: "0.75rem" }}>
                      {[
                        item.solidSubset
                          ? item.solidSubset.length === 1
                            ? `Solids: ${item.solidSubset[0]}`
                            : `Solids: ${item.solidSubset.length}`
                          : `Solids: All`,
                        `Mode: ${item.mode}`
                      ].join(" - ")}
                    </div>
                  </div>
                }
              />
            )}
            noResults={<Menu.Item disabled={true} text="No presets." />}
            onItemSelect={onSelect}
          >
            <Button
              disabled={empty}
              text={label}
              title={empty ? PRESET_PICKER_HINT_TEXT : undefined}
              data-test-id="preset-selector-button"
              rightIcon="caret-down"
            />
          </Select>
        </ShortcutHandler>
      </div>
    );
  },
  isEqual
);

function activateSelect(select: Select<any> | null) {
  if (!select) return;
  // eslint-disable-next-line react/no-find-dom-node
  const selectEl = ReactDOM.findDOMNode(select) as HTMLElement;
  const btnEl = selectEl.querySelector("button");
  if (btnEl) {
    btnEl.click();
  }
}

const PickerContainer = styled.div`
  display: flex;
  justify: space-between;
  align-items: center;
`;

export const CONFIG_PRESETS_QUERY = gql`
  query ConfigPresetsQuery($pipelineName: String!) {
    pipeline(params: { name: $pipelineName }) {
      name
      presets {
        __typename
        name
        mode
        solidSubset
        environmentConfigYaml
      }
      tags {
        key
        value
      }
    }
  }
`;

export const CONFIG_PARTITION_SETS_QUERY = gql`
  query ConfigPartitionSetsQuery($pipelineName: String!) {
    partitionSetsOrError(pipelineName: $pipelineName) {
      __typename
      ... on PartitionSets {
        results {
          name
          mode
          solidSubset
        }
      }
    }
  }
`;

export const CONFIG_PARTITIONS_QUERY = gql`
  query ConfigPartitionsQuery(
    $partitionSetName: String!
    $pipelineName: String!
  ) {
    pipeline(params: { name: $pipelineName }) {
      name
      tags {
        key
        value
      }
    }
    partitionSetOrError(partitionSetName: $partitionSetName) {
      __typename
      ... on PartitionSet {
        partitions {
          results {
            name
            solidSubset
            environmentConfigYaml
            mode
            tags {
              key
              value
            }
          }
        }
      }
    }
  }
`;

function usePresetsAndPartitions(
  pipelineName: string
): {
  presets: Preset[];
  partitionSets: PartitionSet[];
  loading: boolean;
  pipeline?: Pipeline;
} {
  const presetsQuery = useQuery<ConfigPresetsQuery>(CONFIG_PRESETS_QUERY, {
    fetchPolicy: "network-only",
    variables: { pipelineName }
  });
  const partitionSetsQuery = useQuery<ConfigPartitionSetsQuery>(
    CONFIG_PARTITION_SETS_QUERY,
    {
      fetchPolicy: "network-only",
      variables: { pipelineName }
    }
  );

  const byName = (a: { name: string }, b: { name: string }) =>
    a.name.localeCompare(b.name);

  return {
    loading: presetsQuery.loading || partitionSetsQuery.loading,
    presets:
      presetsQuery.data?.pipeline?.__typename === "Pipeline"
        ? presetsQuery.data.pipeline.presets.sort(byName)
        : [],
    partitionSets:
      partitionSetsQuery.data?.partitionSetsOrError?.__typename ===
      "PartitionSets"
        ? partitionSetsQuery.data.partitionSetsOrError.results.sort(byName)
        : [],
    pipeline: presetsQuery.data?.pipeline
  };
}
