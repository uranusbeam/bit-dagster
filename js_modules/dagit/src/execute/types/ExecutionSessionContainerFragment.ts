// @generated
/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: ExecutionSessionContainerFragment
// ====================================================

export interface ExecutionSessionContainerFragment_PipelineNotFoundError {
  __typename: "PipelineNotFoundError" | "PythonError";
}

export interface ExecutionSessionContainerFragment_Pipeline_modes {
  __typename: "Mode";
  name: string;
  description: string | null;
}

export interface ExecutionSessionContainerFragment_Pipeline_tags {
  __typename: "PipelineTag";
  key: string;
  value: string;
}

export interface ExecutionSessionContainerFragment_Pipeline {
  __typename: "Pipeline";
  name: string;
  modes: ExecutionSessionContainerFragment_Pipeline_modes[];
  tags: ExecutionSessionContainerFragment_Pipeline_tags[];
}

export interface ExecutionSessionContainerFragment_InvalidSubsetError_pipeline_modes {
  __typename: "Mode";
  name: string;
  description: string | null;
}

export interface ExecutionSessionContainerFragment_InvalidSubsetError_pipeline_tags {
  __typename: "PipelineTag";
  key: string;
  value: string;
}

export interface ExecutionSessionContainerFragment_InvalidSubsetError_pipeline {
  __typename: "Pipeline";
  name: string;
  modes: ExecutionSessionContainerFragment_InvalidSubsetError_pipeline_modes[];
  tags: ExecutionSessionContainerFragment_InvalidSubsetError_pipeline_tags[];
}

export interface ExecutionSessionContainerFragment_InvalidSubsetError {
  __typename: "InvalidSubsetError";
  message: string;
  pipeline: ExecutionSessionContainerFragment_InvalidSubsetError_pipeline;
}

export type ExecutionSessionContainerFragment = ExecutionSessionContainerFragment_PipelineNotFoundError | ExecutionSessionContainerFragment_Pipeline | ExecutionSessionContainerFragment_InvalidSubsetError;
