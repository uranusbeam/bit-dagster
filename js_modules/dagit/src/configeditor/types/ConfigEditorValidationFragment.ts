// @generated
/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import { EvaluationErrorReason } from "./../../types/globalTypes";

// ====================================================
// GraphQL fragment: ConfigEditorValidationFragment
// ====================================================

export interface ConfigEditorValidationFragment_InvalidSubsetError {
  __typename: "InvalidSubsetError" | "PipelineConfigValidationValid" | "PipelineNotFoundError" | "PythonError";
}

export interface ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries_EvaluationStackPathEntry {
  __typename: "EvaluationStackPathEntry";
  fieldName: string;
}

export interface ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries_EvaluationStackListItemEntry {
  __typename: "EvaluationStackListItemEntry";
  listIndex: number;
}

export type ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries = ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries_EvaluationStackPathEntry | ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries_EvaluationStackListItemEntry;

export interface ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack {
  __typename: "EvaluationStack";
  entries: ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack_entries[];
}

export interface ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors {
  __typename: "FieldNotDefinedConfigError" | "FieldsNotDefinedConfigError" | "MissingFieldConfigError" | "MissingFieldsConfigError" | "RuntimeMismatchConfigError" | "SelectorTypeConfigError";
  reason: EvaluationErrorReason;
  message: string;
  stack: ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors_stack;
}

export interface ConfigEditorValidationFragment_PipelineConfigValidationInvalid {
  __typename: "PipelineConfigValidationInvalid";
  errors: ConfigEditorValidationFragment_PipelineConfigValidationInvalid_errors[];
}

export type ConfigEditorValidationFragment = ConfigEditorValidationFragment_InvalidSubsetError | ConfigEditorValidationFragment_PipelineConfigValidationInvalid;
