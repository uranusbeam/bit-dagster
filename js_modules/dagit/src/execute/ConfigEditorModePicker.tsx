import * as React from "react";
import { Button, Intent, Menu } from "@blueprintjs/core";
import { IconNames } from "@blueprintjs/icons";
import { Select } from "@blueprintjs/select";
import { PipelineDetailsFragment_modes } from "./types/PipelineDetailsFragment";
import { ModeNotFoundError } from "./ExecutionSessionContainer";

interface ConfigEditorModePickerProps {
  modes: PipelineDetailsFragment_modes[];
  modeError?: ModeNotFoundError;
  modeName: string | null;
  onModeChange: (mode: string) => void;
}

interface Mode {
  name: string;
}

const MODE_PICKER_HINT_TEXT = `To add a mode, add a ModeDefinition to the pipeline.`;

const ModeSelect = Select.ofType<Mode>();

export class ConfigEditorModePicker extends React.PureComponent<
  ConfigEditorModePickerProps
> {
  getModeFromProps = (props: ConfigEditorModePickerProps) => {
    return props.modeName
      ? props.modes.find(m => m.name === props.modeName)
      : props.modes[0];
  };

  getCurrentMode = () => {
    return this.getModeFromProps(this.props);
  };

  componentDidMount() {
    const currentMode = this.getCurrentMode();
    if (currentMode) {
      this.props.onModeChange && this.props.onModeChange(currentMode.name);
    }
  }

  componentDidUpdate(prevProps: ConfigEditorModePickerProps) {
    const currentMode = this.getCurrentMode();
    const prevMode = this.getModeFromProps(prevProps);

    if (currentMode && currentMode !== prevMode) {
      this.props.onModeChange && this.props.onModeChange(currentMode.name);
    }
  }

  render() {
    const singleMode = this.props.modes.length === 1;
    const currentMode = this.getCurrentMode();
    const valid = !this.props.modeError;
    const disabled = singleMode && valid;

    return (
      <ModeSelect
        activeItem={currentMode}
        filterable={true}
        disabled={singleMode && valid}
        items={this.props.modes}
        itemPredicate={(query, mode) =>
          query.length === 0 || mode.name.includes(query)
        }
        itemRenderer={(mode, props) => (
          <Menu.Item
            active={props.modifiers.active}
            key={mode.name}
            text={mode.name}
            onClick={props.handleClick}
          />
        )}
        onItemSelect={this.onItemSelect}
      >
        <Button
          icon={valid ? undefined : IconNames.ERROR}
          intent={valid ? Intent.NONE : Intent.WARNING}
          title={disabled ? MODE_PICKER_HINT_TEXT : "Current execution mode"}
          text={
            valid
              ? currentMode
                ? `Mode: ${currentMode.name}`
                : "Select Mode"
              : "Invalid Mode Selection"
          }
          disabled={disabled}
          rightIcon="caret-down"
          data-test-id="mode-picker-button"
        />
      </ModeSelect>
    );
  }

  private onItemSelect = (mode: Mode) => {
    this.props.onModeChange && this.props.onModeChange(mode.name);
  };
}
