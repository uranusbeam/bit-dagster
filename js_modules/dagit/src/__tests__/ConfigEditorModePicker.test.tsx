import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

import { ConfigEditorModePicker } from "../execute/ConfigEditorModePicker";
import { ModeNotFoundError } from "../execute/ExecutionSessionContainer";
import { ExecutionSessionContainerFragment_Pipeline_modes } from "../execute/types/ExecutionSessionContainerFragment";

const defaultMode: ExecutionSessionContainerFragment_Pipeline_modes = {
  __typename: "Mode",
  description: null,
  name: "default"
};

const mode1: ExecutionSessionContainerFragment_Pipeline_modes = {
  __typename: "Mode",
  description: "Mode 1",
  name: "mode_1"
};

const mode2: ExecutionSessionContainerFragment_Pipeline_modes = {
  __typename: "Mode",
  description: "Mode 2",
  name: "mode_2"
};

it("renders single mode pipelines", () => {
  const componentNullSelected = TestRenderer.create(
    <BrowserRouter>
      <ConfigEditorModePicker
        modes={[defaultMode]}
        modeName={null}
        onModeChange={() => null}
      />
    </BrowserRouter>
  );
  expect(componentNullSelected.toJSON()).toMatchSnapshot();

  const componentDefaultSelected = TestRenderer.create(
    <BrowserRouter>
      <ConfigEditorModePicker
        modes={[defaultMode]}
        modeName={"default"}
        onModeChange={() => null}
      />
    </BrowserRouter>
  );
  expect(componentDefaultSelected.toJSON()).toMatchSnapshot();
});

it("renders multi mode pipelines", () => {
  const componentNullSelected = TestRenderer.create(
    <BrowserRouter>
      <ConfigEditorModePicker
        modes={[mode1, mode2]}
        modeName={null}
        onModeChange={() => null}
      />
    </BrowserRouter>
  );
  expect(componentNullSelected.toJSON()).toMatchSnapshot();

  const componentMode1Selected = TestRenderer.create(
    <BrowserRouter>
      <ConfigEditorModePicker
        modes={[mode1, mode2]}
        modeName="mode_1"
        onModeChange={() => null}
      />
    </BrowserRouter>
  );
  expect(componentMode1Selected.toJSON()).toMatchSnapshot();
});

it("renders error mode", () => {
  const error: ModeNotFoundError = {
    __typename: "ModeNotFoundError",
    message: "Mode Not Found"
  };
  const componentNullSelected = TestRenderer.create(
    <BrowserRouter>
      <ConfigEditorModePicker
        modes={[mode1, mode2]}
        modeError={error}
        modeName="mode_1"
        onModeChange={() => null}
      />
    </BrowserRouter>
  );
  expect(componentNullSelected.toJSON()).toMatchSnapshot();
});
