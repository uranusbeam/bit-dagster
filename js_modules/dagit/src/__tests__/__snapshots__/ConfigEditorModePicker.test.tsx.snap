// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error mode 1`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onClick={[Function]}
  >
    <div
      className=""
      onKeyDown={[Function]}
    >
      <button
        className="bp3-button bp3-intent-warning"
        data-test-id="mode-picker-button"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        title="Current execution mode"
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-error"
          icon="error"
        >
          <svg
            data-icon="error"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              error
            </desc>
            <path
              d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm1 13h-2v-2h2v2zm0-3h-2v-7h2v7z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Invalid Mode Selection
        </span>
        <span
          className="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </span>
</span>
`;

exports[`renders multi mode pipelines 1`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onClick={[Function]}
  >
    <div
      className=""
      onKeyDown={[Function]}
    >
      <button
        className="bp3-button"
        data-test-id="mode-picker-button"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        title="Current execution mode"
        type="button"
      >
        <span
          className="bp3-button-text"
        >
          Mode: mode_1
        </span>
        <span
          className="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </span>
</span>
`;

exports[`renders multi mode pipelines 2`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onClick={[Function]}
  >
    <div
      className=""
      onKeyDown={[Function]}
    >
      <button
        className="bp3-button"
        data-test-id="mode-picker-button"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        title="Current execution mode"
        type="button"
      >
        <span
          className="bp3-button-text"
        >
          Mode: mode_1
        </span>
        <span
          className="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </span>
</span>
`;

exports[`renders single mode pipelines 1`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onClick={[Function]}
  >
    <div
      className=""
      onKeyDown={[Function]}
    >
      <button
        className="bp3-button bp3-disabled"
        data-test-id="mode-picker-button"
        disabled={true}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        tabIndex={-1}
        title="To add a mode, add a ModeDefinition to the pipeline."
        type="button"
      >
        <span
          className="bp3-button-text"
        >
          Mode: default
        </span>
        <span
          className="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </span>
</span>
`;

exports[`renders single mode pipelines 2`] = `
<span
  className="bp3-popover-wrapper"
>
  <span
    className="bp3-popover-target"
    onClick={[Function]}
  >
    <div
      className=""
      onKeyDown={[Function]}
    >
      <button
        className="bp3-button bp3-disabled"
        data-test-id="mode-picker-button"
        disabled={true}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        tabIndex={-1}
        title="To add a mode, add a ModeDefinition to the pipeline."
        type="button"
      >
        <span
          className="bp3-button-text"
        >
          Mode: default
        </span>
        <span
          className="bp3-icon bp3-icon-caret-down"
          icon="caret-down"
        >
          <svg
            data-icon="caret-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              caret-down
            </desc>
            <path
              d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </span>
</span>
`;
